pkg/hub/taint/controller.go

```diff
@@ -0,0 +1,98 @@
+	v1 "open-cluster-management.io/api/cluster/v1"
+	"open-cluster-management.io/registration/pkg/helpers"
+)
+
+var (
+	UnavailableTaint = v1.Taint{
+		Key:    v1.ManagedClusterTaintUnavailable,
+		Value:  "",
+		Effect: v1.TaintEffectNoSelect,
+	}
+
+	UnreachableTaint = v1.Taint{
+		Key:    v1.ManagedClusterTaintUnreachable,
+		Value:  "",
+		Effect: v1.TaintEffectNoSelect,
+	}
+)
+
+// taintController
+type taintController struct {
+	clusterClient clientset.Interface
+	clusterLister listerv1.ManagedClusterLister
+	eventRecorder events.Recorder
+}
+
+// NewTaintController creates a new taint controller
+func NewTaintController(
+	clusterClient clientset.Interface,
+	clusterInformer informerv1.ManagedClusterInformer,
+	recorder events.Recorder) factory.Controller {
+	c := &taintController{
+		clusterClient: clusterClient,
+		clusterLister: clusterInformer.Lister(),
+		eventRecorder: recorder.WithComponentSuffix("taint-controller"),
+	}
+	return factory.New().
+		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
+			accessor, _ := meta.Accessor(obj)
+			return accessor.GetName()
+		}, clusterInformer.Informer()).
+		WithSync(c.sync).
+		ToController("taintController", recorder)
+}
+
+func (c *taintController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	managedClusterName := syncCtx.QueueKey()
+	klog.V(4).Infof("Reconciling ManagedCluster %s", managedClusterName)
+	managedCluster, err := c.clusterLister.Get(managedClusterName)
+	if errors.IsNotFound(err) {
+		// Spoke cluster not found, could have been deleted, do nothing.
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	managedCluster = managedCluster.DeepCopy()
+	newTaints := managedCluster.Spec.Taints
+
+	switch {
+	case meta.IsStatusConditionTrue(managedCluster.Status.Conditions, v1.ManagedClusterConditionAvailable):
+		helpers.RemoveTaints(&newTaints, UnavailableTaint, UnreachableTaint)
+		c.eventRecorder.Eventf("ManagedClusterConditionAvailable conditionStatus is True", "remove UnavailableTaint and UnreachableTaint")
+	case meta.IsStatusConditionFalse(managedCluster.Status.Conditions, v1.ManagedClusterConditionAvailable):
+		helpers.RemoveTaints(&newTaints, UnreachableTaint)
+		helpers.AddTaints(&newTaints, UnavailableTaint)
+		c.eventRecorder.Eventf("ManagedClusterConditionAvailable conditionStatus is False", "change UnreachableTaint to UnavailableTaint")
+	case meta.FindStatusCondition(managedCluster.Status.Conditions, v1.ManagedClusterConditionAvailable) == nil ||
+		meta.IsStatusConditionPresentAndEqual(managedCluster.Status.Conditions, v1.ManagedClusterConditionAvailable, metav1.ConditionUnknown):
+		helpers.RemoveTaints(&newTaints, UnavailableTaint)
+		helpers.AddTaints(&newTaints, UnreachableTaint)
+		c.eventRecorder.Eventf("ManagedClusterConditionAvailable conditionStatus is Unknown or without this condition",
+			"change UnavailableTaint to UnreachableTaint")
+	}
+
+	if helpers.IsTaintsEqual(managedCluster.Spec.Taints, newTaints) {
```

Comment:
if `RemoveTaints` and `AddTaints` have a flag, we may eliminate this judgement, am I right?